ISPN-13124 Make SingleFileStore implement NonBlockingStore

danberindei opened the pull request:
https://issues.redhat.com/browse/ISPN-13124

@wburns I did not add a migration from the old "format" using one file per segment. I'm kind of hoping it's not going to be necessary, if SIFS is now the implementation behind `<file-store/>`.

wburns reviewed (commented):
Meant to send this before I left.

wburns commented on core/src/test/java/org/infinispan/persistence/dummy/DummyInMemoryStore.java:
```diff
@@ -281,22 +276,19 @@ private boolean isExpired(MarshallableEntry me, long now) {
-                  int segment = keyPartitioner.getSegment(e.getKey());
-                  return segments.contains(segment);
-               });
+         flowable = Flowable.fromIterable(store.get(0).entrySet());
```
Why was the filter removed? It should be more efficient to filter by segment here and not deserialize then to deserialize every entry and then have it filtered at PersistenceManagerImpl.

danberindei replied:
My thinking was that the store is not going to be segmented when the user doesn't need segments, so the publish methods are going to be invoked mostly with the full set of segments and the segment check is going to be just extra work.

wburns replied:
Just because a store isn't segmented doesn't mean the user doesn't need segments. In many cases this is called with a subset of segments when it is not shared due to topology changes.

But looking closer this change is fine. I forgot that when the store is not segmented we filter out the segments of the keys in the provided filter, so the lines below will handle this anyways. https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java#L740

danberindei replied:
Yeah, I should have been clearer when I said "extra work", I meant "duplicated work" because PMI assumes the store doesn't filter by segment.

And by "the user doesn't need segments" I meant that the user must have set segmented="false" because they know they don't need segments. If they do need segments and they still set segmented="false", then the unmarshalling of the value is just a drop in the bucket compared to all the iteration they have to perform over all their segments, when publishers request only 1 segment at a time.

wburns replied:
> Yeah, I should have been clearer when I said "extra work", I meant "duplicated work" because PMI assumes the store doesn't filter by segment.

It actually does with the provided filter as I mentioned in my previous comment :)

> And by "the user doesn't need segments" I meant that the user must have set segmented="false" because they know they don't need segments. If they do need segments and they still set segmented="false", then the unmarshalling of the value is just a drop in the bucket compared to all the iteration they have to perform over all their segments, when publishers request only 1 segment at a time.

It is definitely _not_ a drop in the bucket. You are talking about deserializing every entry for every segment, no matter what versus only the ones you care about. Iterating and deserializing N entries versus iterating N and deserializing N/2 is a BIG difference (assuming only half the segments are requested).

But as I stated in my previous comment this change is fine as we already filter out the entries in the provided filter automatically for non segmented stores.

danberindei replied:
> You are talking about deserializing every entry for every segment, no matter what versus only the ones you care about.

Sorry I wasn't clear again, but I did not mean that: I was assuming that PMI filters by segment before unmarshalling, if the store is not segmented (because sometimes I trusted the javadoc and sometimes I looked at the PMI code instead, so I wasn't very consistent).

The point I was trying to get to is that a user who cares about performance in a clustered cache will not set segmented="false". If they do disable segmentation for "reasons", my optimization of hashing each key only in the PMI instead of once in the PMI and once in the store won't make a difference in their choice.

wburns replied:
Sounds like we are on the same page then :)

But there is no reason to has in the store as well for this case as the PMI handles the case and it is less code, so removing it 👍

wburns reviewed (commented):
@danberindei I think we need a migration path from previous data. Our internal caches and programmatic usages will still be using SingleFileStore and also those who are still using xml configuration of 12 or older will still have troubles.

wburns commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -1071,53 +1298,107 @@ private void mergeAndLogEntry(FileEntry entry, int mergeCounter) {
+            if (segmentEntries == null)
+               continue;
+
+            List<KeyValuePair<Object, FileEntry>> entriesToPurge = new ArrayList<>();
+            synchronized (segmentEntries) {
+               for (Iterator<Map.Entry<K, FileEntry>> it = segmentEntries.entrySet().iterator(); it.hasNext(); ) {
+                  Map.Entry<K, FileEntry> next = it.next();
+                  FileEntry fe = next.getValue();
+                  if (fe.isExpired(now)) {
+                     it.remove();
+                     entriesToPurge.add(new KeyValuePair<>(next.getKey(), fe));
```
```suggestion
            List<Map.Entry<K, FileEntry>> entriesToPurge = new ArrayList<>();
            synchronized (segmentEntries) {
               for (Iterator<Map.Entry<K, FileEntry>> it = segmentEntries.entrySet().iterator(); it.hasNext(); ) {
                  Map.Entry<K, FileEntry> next = it.next();
                  FileEntry fe = next.getValue();
                  if (fe.isExpired(now)) {
                     it.remove();
                     entriesToPurge.add(next);
```
This way we don't need to allocate the pairs.

danberindei replied:
HashMap's entry iterator gives us the Node that's stored in the table, and I don't think it would be safe to read the value from it without synchronization.

wburns replied:
We hold the read lock, so it shouldn't be possible for the map to change though right?

wburns replied:
The more I think about it, can we not just use the stamped lock to protect the segmentEntries entry as well? This would reduce some code complexity as we don't have to worry about multiple locks when reading or writing it. Or is there a case where it wasn't protecting it?

danberindei replied:
The problem is that we only have 1 `resizeLock` (which was initially used only to synchronize `allocate()` adding to `filePos` w/ `clear()` setting `filePos=0`, as the comment on the field still says), and if we hold it exclusively for every write then we lose the ability of writing to different segments in parallel. Granted, the parallel writes are not the main reason why people were using a segmented SFS before, but it seemed a shame to lose that.

So I only considered going the other way around: using `ConcurrentHashMap` for segment entries so that we can iterate over the entries without synchronization and hopefully avoid some of the large collection allocations we have here and elsewhere. But there I ran into the limitation that we use `LinkedHashMap` to implement `max-entries`. I figured I have a choice between 
1. Use Caffeine for the segment entries, which seemed excessive (but I admit I didn't look too much into it)
2. Keep using a synchronized `LinkedHashMap` when `max-entries` is set, but add a completely separate CHM-based path when `max-entries` is not set
3. (my choice) deprecate `max-entries` and keep using a synchronized `LinkedHashMap`/`HashMap` for now, move to CHM later.

wburns commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -1071,53 +1298,107 @@ private void mergeAndLogEntry(FileEntry entry, int mergeCounter) {
+                  it.set(null);
+
+                  // Already "locked" by being removed from the segmentEntries map
+                  MarshallableEntry<K, V> entry = readFromDisk(fe, next.getKey(), false, false);
```
```suggestion
                  MarshallableEntry<K, V> entry = readFromDisk(fe, next.getKey(), true, true);
```
The goal is that expiration will have values moving forward. This makes me wonder if we should have had an argument to ask for the value on `purgeExpired`.

danberindei replied:
Yeah, a parameter (to be set to true only when an expiration listener has been registered) sounds good.
I thought the value and metadata weren't required because both ClusterExpirationManager and ExpirationManagerImpl expect them to be missing, e.g.
https://github.com/infinispan/infinispan/blob/a40d46817d5f41c74f0f96145d23a8e1e17c6004/core/src/main/java/org/infinispan/expiration/impl/ClusterExpirationManager.java#L439

wburns replied:
That is just the way it worked before, the new SPI was trying to change that so the values would be returned. That way we could finally make Paul happy :)

danberindei replied:
After spending a few good hours debugging a failure in `ExpirationSingleFileStoreDistListenerFunctionalTest.testExpirationOfStoreWhenDataNotInMemory` I finally remembered why I didn't read the value and metadata :)

https://github.com/danberindei/infinispan/blob/edcc01b34c09d991c9b447b47d2d97acfc556e99/core/src/main/java/org/infinispan/interceptors/impl/CallInterceptor.java#L560

The comment is not completely correct: even on the node where purgeExpired did not remove the entry yet, RemoveExpiredCommand has flag SKIP_CACHE_LOAD set, so it can only find the entry in the data container. If the entry does not exist in the data container, the RemoveExpiredCommand fails, and the expiration listener is not invoked, failing the test.

I'll add a comment in the store as well, if I forgot it once I'll surely forget it again!

danberindei replied:
@wburns I added the comment, let me know if you had another fix in mind.

wburns commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -1071,53 +1298,107 @@ private void mergeAndLogEntry(FileEntry entry, int mergeCounter) {
+      if (!segmented && segments.size() < ctx.getCache().getCacheConfiguration().clustering().hash().numSegments()) {
+         return Flowable.fromPublisher(publishKeys(segments, null)).count().toCompletionStage();
+      }
+      return blockingManager.supplyBlocking(() -> blockingApproximateSize(segments), "approximateSize");
```
```suggestion
      return blockingManager.supplyBlocking(() -> blockingApproximateSize(segments), "sfs-approximateSize");
```

wburns commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -1132,6 +1413,75 @@ public SingleFileStoreConfiguration getConfiguration() {
+      if (!segmented) {
+         // Delete keys individually
+         return Flowable.fromPublisher(publishKeys(segments, null))
+                        .doOnNext(k -> delete(0, k))
```
Do we want to provide some back pressure here to prevent blowing up the blocking pool?

danberindei replied:
I forgot that delete() returns a CompletionStage :)

danberindei commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -949,6 +1223,72 @@ public SingleFileStoreConfiguration getConfiguration() {
 
+   @Override
+   public CompletionStage<Void> addSegments(IntSet segments) {
+      if (!segmented) {
```
`PersistenceManagerImpl.shouldInvokeSegmentMethods()` doesn't know anything about the store's configuration, it only checks if the store has the SEGMENTABLE characteristic.

wburns replied:
Which is updated in https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java#L266

danberindei replied:
Sorry, but that looks like a hack to me :) The way I read it, the characteristic is called "segmentable", so its presence means the store could be segmented, but it doesn't have to be segmented.

If SEGMENTABLE really means SEGMENTED, then I guess I could change the `characteristics()` implementation to return SEGMENTABLE only when the configuration is segmented. The NonBlockingStore javadoc says different instances of the store could return different characteristics, as long as the characteristics don't change after start.

If not, then it would be better IMO to keep a separate `segmented` flag in `StoreStatus` and to mention explicitly in the NonBlockingStore javadoc which methods are only called when the configuration is segmented.

wburns replied:
> Sorry, but that looks like a hack to me :) The way I read it, the characteristic is called "segmentable", so its presence means the store could be segmented, but it doesn't have to be segmented.

That is exactly what it does. The store status stores whether it actually is or not at runtime.

> If SEGMENTABLE really means SEGMENTED, then I guess I could change the `characteristics()` implementation to return SEGMENTABLE only when the configuration is segmented.

It doesn't mean that for the store. You are thinking solely from the store perspective. The store status just resuses it for its own purpose is all.

> The NonBlockingStore javadoc says different instances of the store could return different characteristics, as long as the characteristics don't change after start.

Yeah, that is perfectly fine. It works just fine either way.

> If not, then it would be better IMO to keep a separate `segmented` flag in `StoreStatus` and to mention explicitly in the NonBlockingStore javadoc which methods are only called when the configuration is segmented.

The characteristic is that flag on the status. It is already mentioned explicitly on the javadoc for every method. (e.g. https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java#L393)

danberindei replied:
> It doesn't mean that for the store. You are thinking solely from the store perspective. The store status just resuses it for its own purpose is all.

And that's exactly what I'm arguing against: a characteristic SEGMENTABLE shouldn't have one meaning in the store and another meaning in PersistenceManagerImpl.

> It is already mentioned explicitly on the javadoc for every method. (e.g. https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java#L393)

My bad, I should have checked the javadoc for `addSegments()`.

In my defense, the not all the javadocs say the same thing. The `load()`, `delete()`, `write()`, `publishKeys()` and `publishEntries()` javadoc only says "When this is not set the provided {@code segments} parameter may be ignored.", implying that the segments parameter must not be ignored if the store reports that it supports segmentation but segmentation is disabled in the configuration.

And that is indeed how `DummyInMemoryStore` implements them (especially now that I added back the filter in `publishEntries()`).

wburns replied:
> And that's exactly what I'm arguing against: a characteristic SEGMENTABLE shouldn't have one meaning in the store and another meaning in PersistenceManagerImpl.

That's fine, it is personal preference imo. Note that this is done for more than just segmented as well and isn't as clear cut for some.



> In my defense, the not all the javadocs say the same thing. The `load()`, `delete()`, `write()`, `publishKeys()` and `publishEntries()` javadoc only says "When this is not set the provided {@code segments} parameter may be ignored.",

Yeah, because they are different. Note that publishKeys and publishEntries are not invoked if you don't have the BULK_READ characteristic https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java#L649

Same thing for write/delete with READ_ONLY https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java#L359

And same thing for load/contains with WRITE_ONLY
https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java#L286

> implying that the segments parameter must not be ignored if the store reports that it supports segmentation but segmentation is disabled in the configuration.

Sure, that is an oversight and please add in that it can be ignore if the store doesn't have SEGMENTABLE or StoreConfiguration#segmented is false.

> And that is indeed how DummyInMemoryStore implements them (especially now that I added back the filter in publishEntries()).

DO NOT add the filter method back in as I said on the other review comment.

danberindei replied:
> Note that this is done for more than just segmented as well and isn't as clear cut for some.

Again IMO but the fact that it's less clear cut for other characteristics should be even more reason to have two different names for them, to make it clear when we're using one meaning and when we're using the other meaning. Otherwise we're just inviting confusion.

> Sure, that is an oversight and please add in that it can be ignore if the store doesn't have SEGMENTABLE or StoreConfiguration#segmented is false.

:+1: 

> DO NOT add the filter method back in as I said on the other review comment.

I'd rather add it back now that I realized I am using the segments parameter in  `SingleFileStore.publishEntries()` and `RocksDBStore.NonSegmentedRocksDBHandler.publishEntries()` is also using it, but ok :)

danberindei replied:
@wburns I changed `addSegments()` and `removeSegments()` to throw `UnsupportedOperationException`, but FWIW RocksDBTest (which I used as a template for SingleFileStoreTest) still tests that `addSegments` and `removeSegments` work when segmentation is disabled in the configuration. I haven't checked the other stores, but it seems pretty clear I wasn't the only one who didn't read the javadoc of `addSegments` and `removeSegments` :)

danberindei commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -127,7 +140,9 @@
 
    private FileChannel channel;
-   private Map<K, FileEntry> entries;
+   // The list can only grow: entries of removed segments are set to null, not removed
```
The comment was even more outdated than that: the array size doesn't change at all :)

danberindei commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -136,27 +151,47 @@
+      entries = new Map[numSegments];
+      freeList = Collections.synchronizedSortedSet(new TreeSet<>());
+
+      blockingAddSegments(IntSets.immutableRangeSet(numSegments));
```
I'll add a comment, but at this point no other thread can access the store so the lock acquisition won't block.

wburns replied:
I just hope not test does some crazy thing with starting and stopping, but we will find it anyways and this doesn't harm prod running code so 👍

danberindei commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -786,28 +849,41 @@ private FileEntry evict() {
-                  fe.waitUnlocked();
-               for (FileEntry fe : freeList)
-                  fe.waitUnlocked();
+         // Wait until all readers are done reading file entries
```
No, I meant it to apply to the entire `for` statement.
I'll change it to "Wait until all readers are done reading all file entries"

danberindei commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -918,51 +1063,131 @@ public boolean delete(Object key) {
+      }
+      return Flowable.fromIterable(segments)
+                     .concatMap(segment -> {
+                        Flowable<K> segmentPublisher = blockingPublishSegmentKeys(filter, segment);
```
Sure, I can do the same optimization here, I just didn't think publishers are used often enough for it to make a difference.

The optimization only helps in a very particular scenario: passivation with all entries fitting in the data container, so there is no load from disk but there is a load for every cache miss, and a delete for every cache remove.

wburns replied:
My main concern is the first sentence, you are doing non blocking work in a blocking thread and possibly blocking work in a non blocking thread, opposite of what we want.

The optimization is more of a nice to have, I am more worried about the first thing really.

danberindei replied:
I see what you mean now, I was using `blockingManager.blockingPublisher(segmentPublisher)` wrong on the next line because all the potentially blocking work was done in `blockingPublishSegmentKeys()`.

danberindei commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -1071,53 +1298,107 @@ private void mergeAndLogEntry(FileEntry entry, int mergeCounter) {
+      blockingManager.runBlocking(() -> {
+         blockingPurgeExpired(now, processor);
+      }, "sfs-purgeExpired");
+      return Flowable.fromPublisher(processor.serialize());
```
Probably not.
I think at the time I wanted to publish from multiple threads, but I gave up on that (and I should have probably passed `processor.serialize()` to `blockingPurgeExpired()` instead of returning it to the caller, anyway.

wburns replied:
How were you thinking of using multiple threads here? If anything we could add multi threading to the PersistenceManagerImpl to purge stores in parallel maybe, but I see no reason for the SFS to do multi threading here.

danberindei replied:
I was thinking of using one thread to iterate over each segment, all of them publishing entries on the same serialized processor.

danberindei commented:
I finally added the migration from the old file layout with 1 file per segment to the new layout with a single file even when segmented @wburns 

I made a lot of changes with this, so I decided to also squash the "Address review comments"  commit
As I was quashing I realized that I didn't go far enough with the getFirstStore() changes, so I made `getFirstStoreWait()` always return a `WaitDelegatingNonBlockingStore` and I changed `EnsureNonBlockingStore` to extend `WaitDelegatingNonBlockingStore` as well.

wburns commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -127,76 +139,86 @@
+   private void blockingStart() {
+      boolean shouldClear = configuration.purgeOnStartup();
+      boolean readOnly = configuration.ignoreModifications();
+      assert !(shouldClear && readOnly) : "Store can't be configured with both purge and ignore modifications";
```
We should probably have this in a generic spot so that all stores can benefit from it.

danberindei replied:
`AbstractStoreConfigurationBuilder.validateStoreAttributes()` already throws an exception for this combination, I just wanted to be extra sure :)

wburns commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -87,9 +100,8 @@
+public class SingleFileStore<K, V> implements NonBlockingStore<K, V> {
    private static final Log log = LogFactory.getLog(SingleFileStore.class);
 
    public static final byte[] MAGIC_BEFORE_11 = new byte[]{'F', 'C', 'S', '1'}; //<11
```
Shouldn't we add a new MAGIC number for 13 since it migrates data?

danberindei replied:
I thought about it, but then it's exactly the same format as before. 

The data is migrated when the store is segmented, but I don't see a benefit in changing the magic number there either: when you start an older version of Infinispan after the data was migrated to a single file, the old version won't read the migrated file at all, so it won't see the new magic number.

wburns replied:
Doesn't the `migrateNonSegmented` always get invoked though or how does it verify this was completed?

danberindei replied:
If the file exists in the right place and has the latest magic header, `validateExistingFile()` returns `null` and there is no migration.

wburns replied:
Ah sorry I misread `if (!Arrays.equals(MAGIC_LATEST, headerMagic))`

wburns commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -208,47 +230,130 @@ public void start() {
+
+         try (FileChannel newChannel = createNewFile(newFile)) {
+            if (!removeOnly) {
+               copyEntriesFromOldFile(magicHeader, newChannel, channel, file.toString());
```
I believe we are missing the write of the MAGIC header at the beginning of the new file, no?

wburns replied:
Do we also want a flag to say if it is segmented or not? Or that is only in memory representation right?

danberindei replied:
> I believe we are missing the write of the MAGIC header at the beginning of the new file, no?

`createNewFile()` writes the magic header

> Do we also want a flag to say if it is segmented or not? Or that is only in memory representation right?

Right, on disk segmented and non-segmented stores look exactly the same

wburns commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -301,50 +412,171 @@ private PrivateMetadata generateMissingInternalMetadata() {
+   }
+
+   private void copyEntriesFromOldSegmentFiles(FileChannel newChannel) throws Exception {
+      int numSegments = ctx.getCache().getCacheConfiguration().clustering().hash().numSegments();
```
I wonder if we should throw an exception instead if the number of segments don't match the number of files. We are sorta covering the case if we have more segments than before but if we configure less it will miss some data.

danberindei replied:
I thought about it as well, but it's going to get more complicated when the cache is clustered and each node only has a subset of the segments: the migration might succeed or fail depending on exactly which segments were local. I thought it would be simpler if we just recommend keeping the number of segments unchanged.

Does SIFS check the number of segments when migrating data from SFS? Perhaps I could use the same checks.

wburns replied:
> I thought it would be simpler if we just recommend keeping the number of segments unchanged.

Oh I think that is exactly what we should do. Just reading the code however made me think of supporting other configurations. The more I am reading it again that is all this is doing as well. The checks for file not existing just confused me at first, but that can happen for DIST. So we can close this out :)

> Does SIFS check the number of segments when migrating data from SFS? Perhaps I could use the same checks.

This only works with the same number of segments.

wburns reviewed (commented):
Still looking a bit further but wanted to pose these first

wburns commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -1071,12 +1462,70 @@ private void mergeAndLogEntry(FileEntry entry, int mergeCounter) {
+
+               entriesToPurge = collectExpiredEntries(now, segmentEntries);
+            } finally {
+               resizeLock.unlockRead(stamp);
```
Don't we need to hold this read lock until we are done with the FileEntry objects as  before? TBH I am not clear what the locks a protecting anymore sadly :( Or should we lock the FileEntry when we are about to purge them?

danberindei replied:
No, the resizeLock is only used to protect the `entries` array so segments can't be removed while we are iterating them. Each `FileEntry` then has it's own locking, so they are not freed while someone is reading from them.

wburns replied:
> Each FileEntry then has it's own locking, so they are not freed while someone is reading from them.

Yeah, but my second sentence is shouldn't we be doing that? I see it is unlocking the FileEntry in `purgeExpiredEntries` when invoking `readFromDisk` but it is never locked. I was expecting the FileEntry to be locked in `collectExpiredEntries` while holding the resizeLock.

wburns replied:
It looks like you are somewhat, but not completely, getting around it by removing the entry from the ConcurrentMap and the fact that `free` now waits for readers to complete. Unfortunately, there is still a window where a reader can see the FileEntry due to `readFromDisk` decrementing readers without incrementing it in `blockingPurgeExpired`. This means we could decrement a legitimate reader and they may see a FileEntry in a possibly incorrect state as it can be overwritten. I would think this would be a very small window and rarely occur, but we should probably have a JIRA just to not lose it.

danberindei replied:
:+1:

I thought about making the lock release in `readFromDisk` conditional, but I decided it wouldn't be worth the complexity.

danberindei commented:
There was a conflict with another PR, I think the `WriteBehindFaultToleranceTest`. Nothing serious, just an import and a cast, and it should be ok now.

wburns commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -1085,39 +1534,49 @@ public void purge(Executor threadPool, final PurgeListener task) {
+   @Override
+   public CompletionStage<Long> approximateSize(IntSet segments) {
+      if (!segmented && segments.size() < ctx.getCache().getCacheConfiguration().clustering().hash().numSegments()) {
+         return Flowable.fromPublisher(publishKeys(segments, null)).count().toCompletionStage();
```
```suggestion
         return size(segments);
```

danberindei replied:
Yep... I'm looking forward to the day I can remove `maxEntries` and force `segmented=true` all the time :)

wburns commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -918,68 +1221,154 @@ public boolean delete(Object key) {
+      }));
+   }
+
+   public Flowable<MarshallableEntry<K, V>> blockingPublishSegmentEntries(int segment, Predicate<? super K> filter,
```
Doesn't need to be public

wburns commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -1085,39 +1534,49 @@ public void purge(Executor threadPool, final PurgeListener task) {
-                  throw new PersistenceException(e);
-               }
-               if (task != null) task.entryPurged(next.getKey());
+         for (int segment = 0; segment < actualNumSegments; segment++) {
+            Map<K, FileEntry> segmentEntries = getSegmentEntries(segment);
+            if (segmentEntries != null && segments.contains(segment)) {
```
```suggestion
         for (PrimitiveIterator.OfInt iter = segments.iterator(); iter.hasNext(); ) {
            Map<K, FileEntry> segmentEntries = getSegmentEntries(iter.next());
            if (segmentEntries != null) {
```

I don't see why this wouldn't work instead or are you trying to avoid an allocation of the primitive iterator?

danberindei replied:
Yeah, it should work

wburns commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -1071,53 +1463,125 @@ private void mergeAndLogEntry(FileEntry entry, int mergeCounter) {
+            it.set(null);
+
+            // Safe to unlock because the entry was locked collectExpiredEntries
+            // FIXME We cannot load the metadata here or the lifespan check in CallInterceptor will fail
```
This sounds like a bug in the CallInterceptor that you linked. Do you think you can log a JIRA for that as well?

danberindei replied:
I created ISPN-13295 and updated the comment.

I was hoping you'd hit the same problem and fix it in https://github.com/infinispan/infinispan/pull/9524 so I wouldn't have to do anything ;)

wburns replied:
I did not, I am guessing the SFS tests just covers something extra :)

wburns commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -1071,53 +1463,128 @@ private void mergeAndLogEntry(FileEntry entry, int mergeCounter) {
-               if (task != null) task.entryPurged(next.getKey());
-            }
+         if (!segmented) {
+            segments = IntSets.immutableSet(0);
```
```suggestion
            return getSegmentEntries(0).size();
```

But I am fine with the way it is.

danberindei commented:
> I just noticed that a ton of JPA tests seem to reliably fail on your branch for some reason https://ci.infinispan.org/job/Infinispan/job/PR-9407/25/#showFailuresLink Can you confirm what is going on there?

They were related, indeed... there were 2 different issues, both not really related to SingleFileStore
* I removed the default local cache that some tests create just in case, and I missed the JPA tests that were actually using it.
* When I implemented the migration I found a better way (IMO) to handle the wrapping of stores in WaitDelegatingNonBlockingStore (something I had already worked on earlier), but some old usages of `getFirstStore()` needed to change to `getFirstStoreWait()`, and the error is only visible at runtime.

Should be done now, including your other comments.

wburns commented:
Looks like something got messed up in the Async store tests now where it couldn't read the single file store data. https://ci.infinispan.org/job/Infinispan/job/PR-9407/27/testReport/junit/org.infinispan.persistence.support/AsyncStoreWithoutEvictionFunctionalTest/testPutClearPutsegmented_false/

danberindei commented:
The test was calling `SingleFileStore.clear()` and not waiting for the clear to finish before stopping the cache. After looking a bit into the test, it's not doing a very good job of testing the async store wrapper, and where it does it overlaps with AsyncStoreTest (which also doesn't have eviction enabled). So I decided to remove it altogether.

danberindei commented:
After considering it further, I added a commit to make `SingleFileStore.clear()` preserve the header no matter when the store stops/crashes, and to also acquire the resize lock in `stop()` so that it doesn't run concurrently with other writes.

wburns commented on core/src/test/java/org/infinispan/persistence/file/SingleFileStoreTest.java:
```diff
@@ -109,4 +111,21 @@ public void testSegmentsRemovedAndAdded() {
       assertTrue(store.contains(key));
    }
+
+   public void testStopDuringClear() {
```
I am not sure this test provides much as the PersistenceManagerImpl has guarantees to not call stop or start during any other store operation, but no reason for this stop integration.

danberindei replied:
True. I started doing only the truncate(4) bit thinking about crashes, and then I got a bit carried away.

wburns commented:
Integrated into main, thanks @danberindei !
